Bottom: 7b8c858116147d9e92b4317d15fbc574a57c9be2
Top:    0437db70371e2d0cc44614a7ae9461c4dd1a01e0
Author: Peter Clifton <peter@clifton-electronics.co.uk>
Date:   2015-02-14 15:13:39 +0000

Refresh of add-action-to-break-elements

---

diff --git a/src/action.c b/src/action.c
index d3b5aa5..273c8fe 100644
--- a/src/action.c
+++ b/src/action.c
@@ -8035,6 +8035,10 @@ static void SmashElement (ElementType *element)
   Cardinal group;
   LayerType *top_copper, *bottom_copper;
   LayerType *top_silk, *bottom_silk;
+  LayerType *layer;
+  LineType *new_line;
+  ArcType *new_arc;
+  PinType *new_via;
 
   group = GetLayerGroupNumberBySide (SWAP_IDENT ? BOTTOM_SIDE : TOP_SIDE);
   top_copper = &PCB->Data->Layer[PCB->LayerGroups.Entries[group][0]];
@@ -8043,34 +8047,29 @@ static void SmashElement (ElementType *element)
   top_silk = &PCB->Data->Layer[top_silk_layer];
   bottom_silk = &PCB->Data->Layer[bottom_silk_layer];
 
-  /*
-   * At this point the buffer should contain just a single element.
-   * Now we detach the single element from the buffer and then clear the
-   * buffer, ready to receive the smashed elements.  As a result of detaching
-   * it the single element is orphaned from the buffer and thus will not be
-   * free()'d by FreeDataMemory (called via ClearBuffer).  This leaves it
-   * around for us to smash bits off it.  It then becomes our responsibility,
-   * however, to free the single element when we're finished with it.
-   */
-  /* XXX: TODO: Detach element from PCB */
-  //Buffer->Data->Element = NULL;
-  //Buffer->Data->ElementN = 0;
-
   ELEMENTLINE_LOOP (element);
   {
-    CreateNewLineOnLayer (top_silk, /* BOTTOM? */
-                          line->Point1.X, line->Point1.Y,
-                          line->Point2.X, line->Point2.Y,
-                          line->Thickness, 0, NoFlags ());
-    if (line)
-      line->Number = STRDUP (NAMEONPCB_NAME (element));
+    layer = TEST_FLAG (ONSOLDERFLAG, element) ? bottom_silk : top_silk;
+    new_line = CreateNewLineOnLayer (layer,
+                                     line->Point1.X, line->Point1.Y,
+                                     line->Point2.X, line->Point2.Y,
+                                     line->Thickness, 0, NoFlags ());
+    if (new_line)
+      {
+        new_line->Number = STRDUP (NAMEONPCB_NAME (element));
+        AddObjectToCreateUndoList (LINE_TYPE, layer, line, line);
+      }
   }
   END_LOOP;
   ARC_LOOP (element);
   {
-    CreateNewArcOnLayer (top_silk, /* BOTTOM? */
-                         arc->X, arc->Y, arc->Width, arc->Height, arc->StartAngle,
-                         arc->Delta, arc->Thickness, 0, NoFlags ());
+    layer = TEST_FLAG (ONSOLDERFLAG, element) ? bottom_silk : top_silk;
+    new_arc = CreateNewArcOnLayer (layer,
+                                   arc->X, arc->Y, arc->Width, arc->Height, arc->StartAngle,
+                                   arc->Delta, arc->Thickness, 0, NoFlags ());
+    if (new_arc)
+      AddObjectToCreateUndoList (LINE_TYPE, layer, new_line, new_line);
+
   }
   END_LOOP;
   PIN_LOOP (element);
@@ -8080,24 +8079,27 @@ static void SmashElement (ElementType *element)
     if (TEST_FLAG (HOLEFLAG, pin))
       AddFlags (f, HOLEFLAG);
 
-    CreateNewVia (PCB->Data, pin->X, pin->Y,
-                  pin->Thickness, pin->Clearance, pin->Mask,
-                  pin->DrillingHole, pin->Number, f);
+    new_via = CreateNewVia (PCB->Data, pin->X, pin->Y,
+                            pin->Thickness, pin->Clearance, pin->Mask,
+                            pin->DrillingHole, pin->Number, f);
+    if (new_via)
+      AddObjectToCreateUndoList (VIA_TYPE, new_via, new_via, new_via);
   }
   END_LOOP;
   PAD_LOOP (element);
   {
-    LineType *line;
-    line = CreateNewLineOnLayer (TEST_FLAG (ONSOLDERFLAG, pad) ? bottom_copper : top_copper,
-                                 pad->Point1.X, pad->Point1.Y,
-                                 pad->Point2.X, pad->Point2.Y,
-                                 pad->Thickness, pad->Clearance, NoFlags ());
-    if (line)
-      line->Number = STRDUP (pad->Number);
+    layer = TEST_FLAG (ONSOLDERFLAG, pad) ? bottom_copper : top_copper;
+    new_line = CreateNewLineOnLayer (layer,
+                                      pad->Point1.X, pad->Point1.Y,
+                                      pad->Point2.X, pad->Point2.Y,
+                                      pad->Thickness, pad->Clearance, NoFlags ());
+    if (new_line)
+      {
+        new_line->Number = STRDUP (pad->Number);
+        AddObjectToCreateUndoList (LINE_TYPE, layer, new_line, new_line);
+      }
   }
   END_LOOP;
-//  FreeElementMemory (element);
-//  FreeElement (element);
   RemoveElement (element);
 }
